src/datasets/demo_replay.py

```diff
@@ -70,8 +84,63 @@ def create_demo_replay_data(env: BaseEnv, train_tasks: List[Task]) -> Dataset:
         except EnvironmentFailure:
             # We ignore replay data which leads to an environment failure.
             continue
+
+        if nonoptimal_only and _replay_is_optimal(replay_traj, traj, state_idx,
+                                                  oracle_approach, env):
+            continue
+
         if CFG.option_learner != "no_learning":
             for act in replay_traj.actions:
                 act.unset_option()
         replay_dataset.append(replay_traj)
+
+    assert len(replay_dataset) == CFG.offline_data_num_replays
+
     return demo_dataset + replay_dataset
+
+
+def _replay_is_optimal(replay_traj: LowLevelTrajectory,
+                       demo_traj: LowLevelTrajectory, state_idx: int,
+                       oracle_approach: BaseApproach, env: BaseEnv) -> bool:
+    """Plan from the end of the replay to the goal and check whether the result
+    is as good as the demo.
+
+    The state_idx is the index into demo_traj.states where the replay
+    started.
+    """
+    assert demo_traj.states[state_idx].allclose(replay_traj.states[0])
+    # Plan starting at the end of the replay trajectory to the demo goal.
+    task = Task(replay_traj.states[-1], demo_traj.goal)
+    try:
+        policy = oracle_approach.solve(
+            task, timeout=CFG.offline_data_planning_timeout)
+
+        continued_traj, _, solved = utils.run_policy_on_task(
+            policy, task, env.simulate, CFG.max_num_steps_check_policy)
+        assert solved
+    except (ApproachFailure, ApproachTimeout):
+        # If planning fails, assume that the replay was not optimal.
+        return False
+
+    # Get the cost-to-go from the actions. This is not just len(actions)
+    # because we care about the length of the option trajectory, not the
+    # length of the low-level action trajectory.
+    demo_cost_to_go = _actions_to_cost_to_go(demo_traj.actions)
+    replay_actions = demo_traj.actions[:state_idx] + continued_traj.actions
+    # The +1 is for the replay itself, which consists of one option.
+    replay_cost_to_go = 1 + _actions_to_cost_to_go(replay_actions)
+    assert demo_cost_to_go <= replay_cost_to_go, "Demo was not optimal."
+    return demo_cost_to_go == replay_cost_to_go
+
+
+def _actions_to_cost_to_go(actions: Sequence[Action]) -> float:
```

Comment:
we should probably move this to utils and call it `num_options_in_action_sequence` or something?

Reply:
sounds good